### PR TITLE
Feature/dashboard redis queue

### LINKED ISSUE #2981: Monitoring of the redis queues via monit
We should check if we could monitor the redis queues for any backlogging via monit or via our monitoring scripts

### PR DESCRIPTION
**Pull request by satkunas**
# Description
 * Advance the capability of `pfstats`.
   * Add plural (1x or 2x only) JSON Xpaths for appending metric namespace based on results (mirror `type=mysql_query` capability)
   * Add support for multiple JSON Xpath results/rows (mirror `type=mysql_query` capability)
   * Minor `pfstats` bugfixes

# Issue
* Fixes #2981

# Delete branch after merge
YES

# NEWS file entries
* Add API endpoint `/redis/queue/stats`

## Bug Fixes
* Issue with Monitoring of the redis queues via monit (#2981)

### CODE REVIEW

**Comment by julsemaan** [comment]
Make sure you get the milestone to '8.0' for your 8.0 PRs

'+1 patch release' is only for bug fixing that should go in the maintenance branch of the previous releases

File: go/stats/metrics.go [hunk 1]
```diff
@@ -225,3 +256,75 @@ func ProcessMetricConfig(ctx context.Context, conf pfconfigdriver.PfStats) error
...
+		return ret, nil
+
+	default:
+		return nil, errors.New("Unhandled response type")
+	}
+
+	return nil, nil
+}
+
+func Zip(a, b []interface{}) ([][2]interface{}, error) {
```
Comment (julsemaan) [inline go/stats/metrics.go]: This deserves a few comments as to what it does

File: lib/pf/UnifiedApi/Controller/Redis.pm [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        push $queue{$item->{queue}}{outstanding}, { name => $item->{name}, count => $item->{count} };
+    }
+    foreach my $item (@{ $stats->miss_counters }) {
+        push $queue{$item->{queue}}{expired}, { name => $item->{name}, count => $item->{count} };
+    }
+    #build json
+    my $json = [];
+    while( my( $key, $value ) = each %queue ){
+        # rebuild hash,
+        #   replace empty []'s w/ undef
```
Comment (julsemaan) [inline lib/pf/UnifiedApi/Controller/Redis.pm:39]: Random question

Why is it better for these to be undef instead of empty arrays ?

Comment (satkunas) [reply lib/pf/UnifiedApi/Controller/Redis.pm:39]: This is a personal habit for a couple reasons
 * Conditionals are more straight forward on an undef/null than on an empty array
 * It's visually more straight-forward, thus more readable
 * They're easier to struct in go

File: html/pfappserver/root/graph/dashboard.tt [hunk 3]
```diff
@@ -711,5 +712,53 @@ END
...
+        <div class="tab-pane[% IF tab == "queue" %] active[% END %]" id="queue">
+            <div class="card">
+                <div class="card-title">
+                    <h2>[% l('Queue Counters') %]</h2>
+                </div>
+                <div class="card-block">
+                    <div class="row-fluid">
+                        <div class="span18 card">
+                            <div data-netdata="packetfence.redis.queue_stats_count"
+                            data-host="/netdata/127.0.0.1"
```
Comment (fdurand) [inline html/pfappserver/root/graph/dashboard.tt:988]: We need to loop on all cluster members since redis is local to each one.

Comment (satkunas) [reply html/pfappserver/root/graph/dashboard.tt:988]: Should we put in a "cluster" tab (2nd next to system) and put these charts there?

Comment (satkunas) [reply html/pfappserver/root/graph/dashboard.tt:988]: I'm out until next week, unless an emergency is needed. This appears to be a simple front end adjustment in the dashboard template.

Can we merge and then either @cgx or I can update after?

**Comment by julsemaan** [comment]
@satkunas 

I think we will have to postpone this to 8.1 since we can't really afford to have any more features in 8.0 that have a chance of breaking something.

File: lib/pf/UnifiedApi/Controller/Redis.pm (hunk 2 above)
Comment (julsemaan) [reply lib/pf/UnifiedApi/Controller/Redis.pm:39]: The thing is its not in our usual practices to do that so this makes it look different than the rest of our codebase

And in the spec we're "promising" an array and this isn't delivering it.

At least to stay consistent with the rest of our code, this should be an empty array

File: lib/pf/UnifiedApi/Controller/Redis.pm [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        push $json, { queue => $key, stats => {
+            count => $value->{count},
+            expired => ( $value->{expired}->[0] ? $value->{expired} : undef ),
```
Comment (julsemaan) [inline lib/pf/UnifiedApi/Controller/Redis.pm]: Would be more obvious to check if there are elements in it by doing:
```
expired => ( @{$value->{expired}} ? $value->{expired} : undef ),
```

Comment (satkunas) [reply lib/pf/UnifiedApi/Controller/Redis.pm]: I'll defer this change based on your earlier comment: https://github.com/inverse-inc/packetfence/pull/3158#discussion_r183021216

File: lib/pf/UnifiedApi/Controller/Redis.pm (hunk 2 above)
Comment (satkunas) [reply lib/pf/UnifiedApi/Controller/Redis.pm:39]: Changed in https://github.com/inverse-inc/packetfence/pull/3158/commits/06467b8f7d0f0d0d65ff21b17c4be2072faa16bb

**Comment by julsemaan** [comment]
@satkunas 

Can you rebase this against devel to fix the conflicts, then we can merge that

**Review by jrouzierinverse — COMMENTED** [review]
I think it would be better to name the endpoint /api/v1/queues
If we ever change the underlying technology behind we do not have to change name of the endpoint.

File: lib/pf/UnifiedApi.pm [hunk 5]
```diff
@@ -203,6 +203,16 @@ our @API_V1_ROUTES = (
...
+    },
+    {
+        controller => 'Redis',
+        collection => {
+            subroutes    => {
+                'queue/stats' => {
+                    get => 'queue'
+                },
+            },
+        },
```
Comment (jrouzierinverse) [inline lib/pf/UnifiedApi.pm:253]: To avoid the resource end points from being created add 

resource => undef,

**Comment by fdurand** [comment]
Rebase please

**Comment by satkunas** [comment]
@julsemaan @jrouzierinverse @fdurand rebased

File: lib/pf/UnifiedApi.pm (hunk 5 above)
Comment (satkunas) [reply lib/pf/UnifiedApi.pm:253]: added in https://github.com/inverse-inc/packetfence/pull/3158/commits/369898600038d89e2b35cc4e1e854b1fdab9b5ae

**Review by julsemaan — COMMENTED** [review]
I'm happy with the code now, if @jrouzierinverse and @fdurand are happy too then we can merge this

**Comment by extrafu** [comment]
@jrouzierinverse or @fdurand please merge.

**Pull request merged**